Lars update

Saurabh7 opened the pull request:
Not merge ready.

lisitsyn commented on src/shogun/regression/LeastAngleRegressionImpl.h:
```diff
@@ -0,0 +1,51 @@
+namespace shogun
+{
+/** Implementation class for Least Angle Regression */
+class CLeastAngleRegressionImpl
```
Why to even have a class with the only static method?

Saurabh7 replied:
Going to add more.
Wanted to make them public and unit-test.
Can change it back later too.

karlnapf commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -127,6 +131,7 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
 	fill(m_is_active.begin(), m_is_active.end(), false);
 
 	SGVector<float64_t> y = ((CRegressionLabels*) m_labels)->get_labels();
+	Map<VectorXd> map_y(y.vector, y.size());
```
pls dont call them map_* but rather eigen_*

karlnapf commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -174,10 +180,13 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
+		Map<VectorXd> map_corr(&corr[0], n_fea);
+		Map<VectorXd> map_mu(&mu[0], n_vec);
+		
+		map_corr=map_Xy-(map_X.transpose()*map_mu);
```
nice -- much easier to read

vigsterkr replied:
that `map_X.transpose()` = Xorig (the original input)

karlnapf replied:
this is the reason why I want get rid of those lapack calls, they totally hide such things.

vigsterkr replied:
but that lapack thing wasn't even doing this... it was that manual nested loop in the beginning of `train_machine` that created this :P

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.cpp:
```diff
@@ -0,0 +1,83 @@
+	SGMatrix<float64_t> nR(num_active+1, num_active+1);
+	for (int32_t i=0; i < num_active; ++i)
+		for (int32_t j=0; j < num_active; ++j)
+			nR(i,j) = R(i,j);
```
maybe  put a newline here?

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.cpp:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	// R' * R_k = (X' * X)_k = col_k, solving to get R_k
+	Map<VectorXd> col_vec(col_k.vector, col_k.size());
```
These are updates of Cholesky factors. There are eigen3 methods for this...

How they are embedded within pseudo code, an example is here:
https://github.com/jbowlan/pyl1ls/blob/master/lars.py

Eigen3 methods are here:
https://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html#a67b49e986da81c3c90fe1793681e70ad

It would be good if those would be put into linalg (later). But definitely not do this manually ....

karlnapf replied:
rankUpdate

karlnapf commented:
Cool, looks good so far....
Only should use eigen3 for these cholesky updates


Can you give me an update?

Saurabh7 commented:
```
           mlpack    scikit  mlpy    shogun 
diabetes        -  0.040068     -  0.000867 
cosExp          -  0.011672     -  0.001960 
arcene          -  0.305375     -  0.018512 
madelon         -  0.087029     -  0.051801 
mnist           -         -     -  2.312541 
```

vigsterkr commented on src/shogun/regression/LeastAngleRegressionImpl.h:
```diff
@@ -0,0 +1,51 @@
+class CLeastAngleRegressionImpl
+{
+	public:
+		/**   */
```
dont forget the documentation of that method

karlnapf replied:
We also do not want eigen in the interface.
Even if it is impl class, put a SGVector, etc
Creating Map is free

vigsterkr replied:
doxygen is still missing

vigsterkr commented on tests/unit/regression/lars_unittest.cc:
```diff
@@ -228,4 +234,98 @@ TEST(LeastAngleRegression, lars_N_LessThan_D)
+			matn(i,j)=mat(i,j);
+		}
+	}
+	for (index_t i=0 ;i<6; i++)
```
please use variables instead of magical constants like 6 and 4...
so lie define:
```
int32_t mat_rows = 6;
int32_t mat_cols = 6;
...
```
after that you can refer to those variables + changing the unit test from 6 to 60 would be easy :P

vigsterkr commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -138,14 +143,15 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
 		for (int32_t j=0; j < n_fea; ++j)
```
this transposition of the input matrix should be handled by eigen or linalg... doing this with a double nested loops is soo... 101 computer science

vigsterkr commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -138,14 +143,15 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
 	// Xy = X' * y
-	cblas_dgemv(CblasColMajor, CblasTrans, n_vec, n_fea, 1, X.matrix, n_vec,
-		y.vector, 1, 0, &Xy[0], 1);
+	map_Xy=map_X.transpose()*map_y;
```
so here basically we X^T^T, i.e. double transpose the original input..... so we could just use the input itself right? :)

vigsterkr replied:
shouldn't this be then `map_Xr*map_y`?

Saurabh7 commented:
Tried to make similar:
https://gist.github.com/Saurabh7/b07c65df21a4fe498284a8086fe617b8

Saurabh7 commented:
Ok, I have updated almost all lapack calls, also using eigen's column accessors now everywhere.
Parallelized over loops where dot products are needed (since cant slice matrices).

I tried different combination of paramters for scikit and shogun extensively. 
Also changed Shogun's early stopping criterion to `max_corr/n_vec > CMath::MACHINE_EPSILON`  (similar to scikit btw).   But Shogun still goes for the full path unnecessarily sometimes.

Saurabh7 commented:
Changed the loops for transpose. If we want to remove it completely will have to specify row major eigen matrices since further we have to loop over features (which are cols right now) multiple times.

vigsterkr commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -187,7 +181,7 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
 		Map<VectorXd> map_mu(&mu[0], n_vec);
 		
 		map_corr=map_Xy-(map_X.transpose()*map_mu);
```
why don't use just have 2 maps?
`map_X` and `map_Xt`

karlnapf replied:
+1

vigsterkr commented on src/shogun/regression/LeastAngleRegressionImpl.cpp:
```diff
@@ -40,23 +40,23 @@ using namespace Eigen;
 namespace shogun
 {
-SGMatrix<float64_t> CLeastAngleRegressionImpl::cholesky_insert_helper(SGMatrix<float64_t> X, SGMatrix<float64_t> R, VectorXd x_vec, SGVector<float64_t> active_set)
+SGMatrix<float64_t> CLeastAngleRegressionImpl::cholesky_insert_helper(Ref<MatrixXd> map_X, SGMatrix<float64_t>& R, int32_t i_max_corr, SGVector<float64_t> active_set)
```
can't we just pass around SGVector/SGMatrix instead of going full into eigen?

Saurabh7 replied:
Then I have to transpose the matrix inside.

karlnapf replied:
why not create a Map inside and transpose it?

Saurabh7 replied:
Leads to additional overhead if I save the transpose. If I use `map.transpose().col(i)` too many cache misses. So options might be pass Eigen things or manually tranpose SGMatrix initally or any suggestions ?

Saurabh7 replied:
Or mayb use directly, no helper method.

vigsterkr replied:
@Saurabh7 maybe this is the reason why in the initial implementation there was that transposition of the input matrix. but it'd be good to actually use a more optimal solution for matrix transposition than nested loops.
all lapack/blas functions has `TRANS` function arg that will let you do the operation on the original or the transposed version of the matrix. would be interesting to see the cache misses there ... :)

Saurabh7 replied:
Develop version: [output](https://gist.github.com/Saurabh7/65c920f3813908358ded0c4ffde5b38b)
where lapack routines are used.
Other than that we need the transpose to extract columns and take ddots

Saurabh7 commented:
Ok when I avoid the transpose at the start and use `map.transpose().col(i)` , this leads to larger runtimes.
https://gist.github.com/Saurabh7/ab36961c51cd5e69213096527b4d2810

karlnapf commented:
Interesting.

The only reason I can think of is that this might be due to cache effects and the prefetcher.
Can you maybe check that hypothesis?
If so, then the work of the double transpose might is worth it

karlnapf commented:
It is important that you document these things, or even put them in the
benchmarking framework.
For later.

What are the current results?

2016-06-08 12:56 GMT+01:00 Saurabh Mahindre <notifications@github.com>:

> Ok, I have updated almost all lapack calls, also using eigen's column
> accessors now everywhere.
> Parallelized over loops where dot products are needed (since cant slice
> matrices).
>
> I tried different combination of paramters for scikit and shogun
> extensively.
> Also changed Shogun's early stopping criterion to max_corr/n_vec >
> CMath::MACHINE_EPSILON (similar to scikit btw). But Shogun still goes for
> the full path unnecessarily sometimes.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/3243#issuecomment-224567574>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAqqv4TXQAo5HsB3CCt7JakpsPFdX7xQks5qJq3dgaJpZM4ItWuD>
> .
>

karlnapf commented:
oh also, *when* does shognu do the full path unnecessarily? Would be good
to identify these cases

2016-06-08 12:56 GMT+01:00 Saurabh Mahindre <notifications@github.com>:

> Ok, I have updated almost all lapack calls, also using eigen's column
> accessors now everywhere.
> Parallelized over loops where dot products are needed (since cant slice
> matrices).
>
> I tried different combination of paramters for scikit and shogun
> extensively.
> Also changed Shogun's early stopping criterion to max_corr/n_vec >
> CMath::MACHINE_EPSILON (similar to scikit btw). But Shogun still goes for
> the full path unnecessarily sometimes.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/3243#issuecomment-224567574>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAqqv4TXQAo5HsB3CCt7JakpsPFdX7xQks5qJq3dgaJpZM4ItWuD>
> .
>

Saurabh7 commented:
with current version: https://gist.github.com/Saurabh7/d2d4ed03b7433f8c33459b106bdf5375

with `map.transpose().col(i)`: https://gist.github.com/Saurabh7/da64a4fcecbc78dcf6fd8f17b5fd80a4

Saurabh7 commented:
Shogun's epsilon was very small ` CMath::MACHINE_EPSILON` which lead to continuing the iters even when max corelation was very low.
Scikit uses eps as `np.finfo(np.float32).eps` which comes out to be 1.1920929e-07 here.
I used this for now.

some values
 ```
  N     D     SG time    SK time       SG mse       SK mse    SG iters    SK iters
----  ----  ----------  ---------  -----------  -----------  ----------  ----------
2000   100  0.00801992  0.0135689  9.25081e+07  9.25081e+07          35          33
2000   500  0.339843    0.240674   2.52726e+09  2.52726e+09         167         165
2000  1000  1.5217      7.07059    8.8488e+09   9.34494e+09         352        1024
5000  1000  3.13794     2.2291     2.3638e+10   2.3638e+10          333         331
5000  2000  18.5837     68.6767    1.04143e+11  1.05728e+11         679        2026
```

vigsterkr commented:
@Saurabh7 thnx for the benchmarks... any idea(s) why that diff between SG and SK in case of `N=5000` and `D=1000`? or that's the cost of that two extra iteration?

Saurabh7 commented:
OK heres comparison of latest version for different number of relevant features:
(setting max iters for scikit at some places as it fails to stop)

NUM FEATS * 0.8 : https://gist.github.com/Saurabh7/f0258a66b331728ce995bc96d0f528f7
NUM FEATS* 0.5 : https://gist.github.com/Saurabh7/5ebd29f17313a335ced279db0a999b1e
NUM FEATS* 0.3: https://gist.github.com/Saurabh7/45b2a1ee0524cdda88e6af405c849e3e

vigsterkr commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -127,25 +130,26 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
+	Map<MatrixXd> map_X(X.matrix, n_vec, n_fea);	
+   	map_X = map_Xr.transpose();
+	
+	float64_t* X_active = SG_MALLOC(float64_t, n_vec*n_fea);
```
this sounds like a `SGVector<float64_t>` no?

karlnapf replied:
yes

vigsterkr commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -172,80 +176,86 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
 	//========================================
 	int32_t nloop=0;
-	while (m_num_active < max_active_allowed && max_corr > CMath::MACHINE_EPSILON && !stop_cond)
+	while (m_num_active < max_active_allowed && max_corr/n_vec > 1.192E-07 && !stop_cond)
```
mmm use a macro or a static const for magical constant values

karlnapf replied:
this should be a parameter

vigsterkr commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -172,80 +176,86 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
+		Map<VectorXd> map_corr(&corr[0], n_fea);
+		Map<VectorXd> map_mu(&mu[0], n_vec);
+		
+		map_corr = map_Xy - (map_X.transpose()*map_mu);
```
again why not to use `map_Xr*map_mu` here instead of `map_X.transpose()*map_mu`?

vigsterkr commented on src/shogun/regression/LeastAngleRegression.h:
```diff
@@ -16,6 +16,7 @@
 #ifdef HAVE_LAPACK
 #include <vector>
 #include <shogun/machine/LinearMachine.h>
+#include <shogun/mathematics/eigen3.h>
```
do we really need this?

karlnapf commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -172,80 +176,86 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
 		if (m_num_active < n_fea)
 		{
+			#pragma omp parallel for shared(gamma)
 			for (int32_t i=0; i < n_fea; ++i)
```
this parallel helps?

Saurabh7 replied:
Havent checked exactly how much it helps. Its just parallelizing the dot products.

karlnapf commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -127,25 +130,26 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
+	SGMatrix<float64_t> X (n_vec, n_fea);
+	Map<MatrixXd> map_Xr(Xorig.matrix, n_fea, n_vec);
+	Map<MatrixXd> map_X(X.matrix, n_vec, n_fea);	
+   	map_X = map_Xr.transpose();
```
you you either use spaces or tab stops (minor)

vigsterkr commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -127,25 +130,26 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
 
 	SGVector<float64_t> y = ((CRegressionLabels*) m_labels)->get_labels();
+	Map<VectorXd> map_y(y.vector, y.size());
 	SGMatrix<float64_t> Xorig = feats->get_feature_matrix();
```
any reason for example why LARS shouldn't be able to work on `SGMatrix<float32_t>` ?
but yeah this should be addressed in a consequent PR ;)

karlnapf commented on src/shogun/regression/LeastAngleRegression.cpp:
```diff
@@ -127,25 +130,26 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
+	float64_t* X_active = SG_MALLOC(float64_t, n_vec*n_fea);
 
 	// beta is the estimator
 	vector<float64_t> beta = make_vector(n_fea, 0);
```
why using vector here?

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.cpp:
```diff
@@ -0,0 +1,69 @@
+
+#include <shogun/mathematics/Math.h>
+#include <shogun/regression/LeastAngleRegressionImpl.h>
+#include <shogun/mathematics/eigen3.h>
```
this line has to go

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.cpp:
```diff
@@ -0,0 +1,69 @@
+	Map<MatrixXd> map_R_new(R_new.matrix, R_new.num_rows, R_new.num_cols);
+
+	map_R_new.block(0, 0, num_active, num_active) = map_R;
+	for (int32_t i=0; i < num_active; ++i)
```
you can do this with memset

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.cpp:
```diff
@@ -0,0 +1,69 @@
+	map_R_new.block(0, 0, num_active, num_active) = map_R;
+	for (int32_t i=0; i < num_active; ++i)
+		map_R_new(i, num_active) = R_k[i];
+	for (int32_t i=0; i < num_active; ++i)
```
same here

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.cpp:
```diff
@@ -0,0 +1,69 @@
+
+namespace shogun
+{
+SGMatrix<float64_t> CLeastAngleRegressionImpl::cholesky_insert_helper(SGMatrix<float64_t>& X, float64_t* X_active, SGMatrix<float64_t>& R, int32_t i_max_corr , int32_t num_active)
```
some of the matrices can be const references

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.h:
```diff
@@ -0,0 +1,51 @@
+#include <vector>
+#include <shogun/lib/common.h>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/mathematics/eigen3.h>
```
please no eigen3 includes in .h files (even if this is impl)

karlnapf commented on src/shogun/regression/LeastAngleRegressionImpl.h:
```diff
@@ -0,0 +1,51 @@
+#define _LARSIMPL_H_
+
+#include <shogun/lib/config.h>
+#include <vector>
```
this one should go as well

Saurabh7 commented:
Updated and squashed .

vigsterkr commented:
well, it's time to merge! :)

karlnapf commented on tests/unit/regression/lars_unittest.cc:
```diff
@@ -228,4 +233,113 @@ TEST(LeastAngleRegression, lars_N_LessThan_D)
+
+TEST(LeastAngleRegression, cholesky_insert)
+{
+	class lars_helper: public CLeastAngleRegression
```
usually, these are called "mock" or so, but ok

karlnapf commented:
Great that this is merged :)

karlnapf commented:
Now it is time for the final benchmarks! Looking forward to those
